File: 160. Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        //set will tell me if I have intersected
        ListNode* current = headA;
        set<ListNode*> visited;
        
        while(current != nullptr){
            visited.insert(current);
            current = current->next;
        }

        current = headB;
        while(current != nullptr){
            if(visited.contains(current)) return current; 
            current = current->next;
        }
        return nullptr;

        
    }
};

File: 160. Intersection of Two Linked Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode currA = headA, currB = headB;
        int lenA = 0, lenB = 0;
        while(currA != null){
            lenA++;
            currA = currA.next;
        }
        while(currB != null){
            lenB++;
            currB = currB.next;
        }
        currA = headA;
        currB = headB;
        if(lenA > lenB){
            int diff = lenA - lenB;
            while(diff != 0){ 
                currA = currA.next;
                diff--;
            }
        }
        else if(lenB > lenA){
            int diff = lenB - lenA;
            while(diff != 0){ 
                currB = currB.next;
                diff--;
            }
        }


        while(currA != null && currB != null){
            if(currA == currB) return currA;
            currA = currA.next;
            currB = currB.next;
        }

        return null;
    }
}